ambry-store/src/main/java/com/github/ambry/store/HybridCompactionPolicy.java

```diff
@@ -0,0 +1,192 @@
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.codehaus.jackson.map.ObjectMapper;
+
+import static com.github.ambry.utils.Utils.*;
+
+
+/**
+ * HybridCompactionPolicy will run compaction with StatsBasedCompactionPolicy more frequently and with CompactAllPolicy
+ * one time out of storeCompactionPolicySwitchPeriod.
+ */
+public class HybridCompactionPolicy implements CompactionPolicy {
+  private final Time time;
+  private final StoreConfig storeConfig;
+  private static final Logger logger = LoggerFactory.getLogger(HybridCompactionPolicy.class);
+  private final Map<String, CompactionPolicySwitchInfo> blobToCompactionPolicySwitchInfoMap;
+  private final ObjectMapper mapper = new ObjectMapper();
+  private static final String COMPACTION_POLICY_COUNTER = "compactionPolicyCounter";
+  private static final String VALUE = "value";
+  private static final String LAST_COMPACT_ALL_TIME = "lastCompactAllTime";
+  private static final String COMPACT_POLICY_INFO_PATH = "/compactionPolicyInfo.";
+  private static final int INIT_COMPACT_ALL_TIME = 0;
+  private static final int INIT_COUNTER_VALUE = 0;
+
+  HybridCompactionPolicy(StoreConfig storeConfig, Time time) {
+    this.storeConfig = storeConfig;
+    this.time = time;
+    this.blobToCompactionPolicySwitchInfoMap = new HashMap<>();
+  }
+
+  /**
+   * @param totalCapacity Total capacity of the {@link BlobStore}
+   * @param usedCapacity Used capacity of the {@link BlobStore}
+   * @param segmentCapacity Segment capacity of a {@link LogSegment}
+   * @param segmentHeaderSize Segment header size of a {@link LogSegment}
+   * @param logSegmentsNotInJournal {@link List<String> } of log segment names which has non overlapping entries with
+   *                                {@link Journal}
+   * @param blobStoreStats {@link BlobStoreStats} pertaining to the {@link BlobStore} for which
+   * {@link CompactionDetails} are requested
+   * @param dataDir the dir to store {@link CompactionPolicySwitchInfo}
+   * @return {@link CompactAllPolicy} or {@link StatsBasedCompactionPolicy}'s {@link CompactionDetails} depends on the switching rules.
+   * @throws StoreException
+   */
+  @Override
+  public CompactionDetails getCompactionDetails(long totalCapacity, long usedCapacity, long segmentCapacity,
+      long segmentHeaderSize, List<String> logSegmentsNotInJournal, BlobStoreStats blobStoreStats, String dataDir)
+      throws StoreException {
+    String storeId = blobStoreStats.getStoreId();
+    CompactionPolicySwitchInfo compactionPolicySwitchInfo = blobToCompactionPolicySwitchInfoMap.getOrDefault(storeId,
+        new CompactionPolicySwitchInfo(new CompactionPolicyCounter(storeConfig), INIT_COMPACT_ALL_TIME));
+    blobToCompactionPolicySwitchInfoMap.put(storeId, compactionPolicySwitchInfo);
+    CompactionPolicy selectCompactionPolicy =
+        selectCompactionPolicyAndUpdateCompactionPolicySwitchInfo(dataDir, storeId, compactionPolicySwitchInfo);
```

Comment:
we can make logic a bit better without checking file every time selecting a policy. Something like
```
if (!blobToCompactionPolicySwitchInfos.contain(storeId)) {
    File file = new File(Paths.get(dataDir, COMPACT_POLICY_INFO_PATH))
    if (file.exists()) {
        ObjectMapper objectMapper = new ObjectMapper();
        CompactionPolicySwitchInfo switchInfo = objectMapper.readValue(file, CompactionPolicySwithInfo.class);
        blobToCompactionPolicySwtichInfos.put(storeId, switchInfo);
    } else {
        blobToCompactionPolicySwtichInfos.put(storeId, new CompactionPolicySwitchInfo(new CompactionPolicyCounter(storeConfig), INIT_COMPACT_ALL_TIME));
    }
}
```

Reply:
Updated.